libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -135,50 +214,58 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 
 void ContractLevelChecker::checkIllegalOverrides(ContractDefinition const& _contract)
 {
-	// TODO unify this at a later point. for this we need to put the constness and the access specifier
-	// into the types
-	map<string, vector<FunctionDefinition const*>> functions;
-	map<string, ModifierDefinition const*> modifiers;
+	FunctionSet const& funcSet = getBaseFunctions(&_contract);
+	ModifierSet const& modSet = getBaseModifiers(&_contract);
 
-	// We search from derived to base, so the stored item causes the error.
-	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+	checkModifierOverrides(funcSet, modSet, _contract.functionModifiers());
+
+	for (FunctionDefinition const* function: _contract.definedFunctions())
 	{
-		for (FunctionDefinition const* function: contract->definedFunctions())
-		{
-			if (function->isConstructor())
-				continue; // constructors can neither be overridden nor override anything
-			string const& name = function->name();
-			if (modifiers.count(name))
-				m_errorReporter.typeError(modifiers[name]->location(), "Override changes function to modifier.");
+		if (contains_if(modSet, MatchByName{function->name()}))
+			m_errorReporter.typeError(function->location(), "Override changes modifier to function.");
```

Comment:
This error is reported even if the modifier is `private` in a base class. I have the impression that this is a change in behaviour. Do we have tests for that?

Reply:
The previous code used
```
		for (ModifierDefinition const* modifier: contract->functionModifiers())
		{
			string const& name = modifier->name();
			ModifierDefinition const*& override = modifiers[name];
			if (!override)
				override = modifier;
			else if (ModifierType(*override) != ModifierType(*modifier))
				m_errorReporter.typeError(override->location(), "Override changes modifier signature.");
			if (!functions[name].empty())
				m_errorReporter.typeError(override->location(), "Override changes modifier to function.");
		}
```

It uses the same function to acquire the modifiers and it doesn't do any kind of visibility check.. so that behavior looks identical to me. 

After some testing and code investigation: I don't think modifiers can have any visibility specified. The parser rejects it if you specify anything.

Reply:
Ah ok, good then!